src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs

```diff
@@ -6582,11 +6588,104 @@ static void copyRelevantAddMethodDiagnostics(BindingDiagnosticBag source, Bindin
+            Debug.Assert(ConversionsBase.IsKeyValuePairType(Compilation, expr.Type, WellKnownType.System_Collections_Generic_KeyValuePair_KV, out _, out _));
+
+            if (getKeyMethod is null ||
+                getValueMethod is null)
+            {
+                return new BoundBadExpression(syntax, LookupResultKind.Empty, [], [expr], type: CreateErrorType());
+            }
+
+            var exprType = (NamedTypeSymbol)expr.Type;
+            return BindDictionaryItemAssignment(
+                syntax,
+                implicitReceiver,
+                setMethod,
+                CreateConversion(
+                    bindKeyOrValue(syntax, expr, getKeyMethod.AsMember(exprType)),
+                    keyConversion,
+                    destinationKeyType,
+                    diagnostics),
+                CreateConversion(
+                    bindKeyOrValue(syntax, expr, getValueMethod.AsMember(exprType)),
+                    valueConversion,
+                    destinationValueType,
+                    diagnostics));
+
+            static BoundCall bindKeyOrValue(SyntaxNode syntax, BoundExpression expr, MethodSymbol getMethod)
+            {
+                return new BoundCall(
+                    syntax,
+                    receiverOpt: expr,
+                    // PROTOTYPE: What is the correct value for receiver cloning? Test with expression element
+                    // and spread element where the item is ref KeyValuePair<,> or ref readonly KeyValuePair<,>.
+                    initialBindingReceiverIsSubjectToCloning: ThreeState.False,
+                    method: getMethod,
+                    arguments: [],
+                    argumentNamesOpt: default,
+                    argumentRefKindsOpt: default,
+                    isDelegateCall: false,
+                    expanded: false,
+                    invokedAsExtensionMethod: false,
+                    argsToParamsOpt: default,
+                    defaultArguments: default,
+                    resultKind: LookupResultKind.Viable,
+                    type: getMethod.ReturnType).MakeCompilerGenerated();
+            }
+        }
+
+        private BoundExpression BindDictionaryItemAssignment(
+            SyntaxNode syntax,
+            BoundObjectOrCollectionValuePlaceholder implicitReceiver,
+            MethodSymbol? setMethod,
+            BoundExpression key,
+            BoundExpression value)
+        {
+            if (setMethod is null)
+            {
+                return new BoundBadExpression(syntax, LookupResultKind.Empty, [], [key, value], type: CreateErrorType());
+            }
+
+            // key and value should have been converted by the caller.
+            Debug.Assert(ConversionsBase.HasIdentityConversion(key.Type, setMethod.Parameters[0].Type));
+            Debug.Assert(ConversionsBase.HasIdentityConversion(value.Type, setMethod.Parameters[1].Type));
+            return new BoundCall(
```

Comment:
I'm not sure that I can articulate a strong reason against this design, but I expected we'd produce a BoundIndexerAccess with two BoundPropertyAccess nodes in it

Reply:
Added a PROTOTYPE comment where the accessor is obtained from `GetCollectionExpressionApplicableIndexer()`.